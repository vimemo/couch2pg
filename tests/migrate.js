import migrate from '../lib/pgmigration'
import db, {ensureDatabaseExists} from '../lib/pgconnection'

const PG_URL = 'postgres://localhost:5432/migration-test'
const pg = db(PG_URL)

describe('migration', () => {
  const cleanUp = async () => {
    await pg.schema.dropTableIfExists('couchdb')
    await pg.schema.dropTableIfExists('couchdb_progress')
    await pg.schema.dropTableIfExists('couch2pg_migrations')
  }

  beforeEach(async () => {
    try {
      await cleanUp()
    } catch(err) {
      if(!err.message.includes('does not exist')){//db does not exist
        throw err
      }
    }
    await ensureDatabaseExists(PG_URL)
  })
  afterEach(() => cleanUp())
  afterAll(() => {
    try {
      pg.destroy()
    } catch(err) {
      if(!err.message.includes('does not exist')){//db does not exist
        throw err
      }
    }
  })

  test('migration creates tables and indexes', async () => {
    const migrations = await migrate(PG_URL)
    expect(migrations).toMatchSnapshot()
    expect(await pg.schema.hasTable('couchdb')).toBe(true)
    expect(await pg.schema.hasTable('couchdb_progress')).toBe(true)
    expect(await pg.schema.hasTable('couch2pg_migrations')).toBe(true)

    const INDEXES_QUERY = 'select * from pg_indexes where tablename not like \'pg%\''
    const indexes = (await pg.schema.raw(INDEXES_QUERY)).rows
    expect(indexes[0].indexname).toBe('couch2pg_migrations_pkey')
    expect(indexes[1].indexname).toBe('couchdb_doc_type')
    expect(indexes[2].indexname).toBe('couchdb_doc_uuid')

    const seqs = (await pg.raw('select * from couchdb_progress')).rows
    expect(seqs.length).toBe(0)
  })

  test('migration to an existing seq without source', async () => {
    const conn = db(PG_URL)
    await conn.raw('CREATE TABLE couchdb_progress(seq varchar)')
    await conn('couchdb_progress').insert({seq:'44'})
    await migrate(PG_URL)
    const seqs = (await pg.raw('select * from couchdb_progress')).rows
    expect(seqs.length).toBe(1)
    expect(seqs[0].seq).toBe('44')
    expect(seqs[0].source).toEqual('default-source')
  })
})
