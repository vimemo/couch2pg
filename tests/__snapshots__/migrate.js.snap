// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migration migration creates tables and indexes 1`] = `
Array [
  Object {
    "action": "do",
    "filename": "201606091247.do.2318-standardise-schema.sql",
    "getSql": [Function],
    "md5": "80caa5bfed7af40fca8f5b0d2321b96d",
    "name": "2318-standardise-schema",
    "version": 201606091247,
  },
  Object {
    "action": "do",
    "filename": "201606091248.do.2124-support-edit-and-delete-of-documents.sql",
    "getSql": [Function],
    "md5": "1e4b99477f52e456fd4876b1986ec68a",
    "name": "2124-support-edit-and-delete-of-documents",
    "version": 201606091248,
  },
  Object {
    "action": "do",
    "filename": "201607132040.do.2506-unique-id-constraint.sql",
    "getSql": [Function],
    "md5": "6e48e48a083a4a06029fe0b231d4f3b1",
    "name": "2506-unique-id-constraint",
    "version": 201607132040,
  },
  Object {
    "action": "do",
    "filename": "201607210832.do.2523-remove-unused-index.sql",
    "getSql": [Function],
    "md5": "dd4854328c111126b3418256cbbbbdae",
    "name": "2523-remove-unused-index",
    "version": 201607210832,
  },
  Object {
    "action": "do",
    "filename": "201611271809.do.2923-couch2-support.sql",
    "getSql": [Function],
    "md5": "1e35b1bd14c9c04e186128d45b3163be",
    "name": "2923-couch2-support",
    "version": 201611271809,
  },
  Object {
    "action": "do",
    "filename": "201803202020.do.3423-multiple-db-support.sql",
    "getSql": [Function],
    "md5": "42b98086b5f07c327655612f575e01e8",
    "name": "3423-multiple-db-support",
    "version": 201803202020,
  },
]
`;
